### PR TITLE
[SPARK-35051][SQL] Support add/subtract of a day-time interval to/from a date

### PR DESCRIPTION
**Pull request by MaxGekk**
### What changes were proposed in this pull request?
Support `date +/- day-time interval`. In the PR, I propose to update the binary arithmetic rules, and cast an input date to a timestamp at the session time zone, and then add a day-time interval to it.

### Why are the changes needed?
1. To conform the ANSI SQL standard which requires to support such operation over dates and intervals:
Screenshot 2021-03-12 at 11 36 14
2. To fix the regression comparing to the recent Spark release 3.1 with default settings.

Before the changes:
```sql
spark-sql> select date'now' + (timestamp'now' - timestamp'yesterday');
Error in query: cannot resolve 'DATE '2021-04-14' + subtracttimestamps(TIMESTAMP '2021-04-14 18:14:56.497', TIMESTAMP '2021-04-13 00:00:00')' due to data type mismatch: argument 1 requires timestamp type, however, 'DATE '2021-04-14'' is of date type.; line 1 pos 7;
'Project [unresolvedalias(cast(2021-04-14 + subtracttimestamps(2021-04-14 18:14:56.497, 2021-04-13 00:00:00, false, Some(Europe/Moscow)) as date), None)]
+- OneRowRelation
```

Spark 3.1:
```sql
spark-sql> select date'now' + (timestamp'now' - timestamp'yesterday');
2021-04-15
```

Hive:
```sql
0: jdbc:hive2://localhost:10000/default> select date'2021-04-14' + (timestamp'2020-04-14 18:15:30' - timestamp'2020-04-13 00:00:00');
+------------------------+
|          _c0           |
+------------------------+
| 2021-04-15 18:15:30.0  |
+------------------------+
```

### Does this PR introduce _any_ user-facing change?
Should not since new intervals have not been released yet.

After the changes:
```sql
spark-sql> select date'now' + (timestamp'now' - timestamp'yesterday');
2021-04-15 18:13:16.555
```

### How was this patch tested?
By running new tests:
```
$ build/sbt "test:testOnly *ColumnExpressionSuite"
```

### CODE REVIEW

**Comment by cloud-fan** [comment]
> Spark 3.1:
> spark-sql> select date'now' + (timestamp'now' - timestamp'yesterday');
> 2021-04-15

Spark 3.1 truncates timestamp to date? I think the new behavior makes more sense but shall we add an item in the migration guide?

**Comment by MaxGekk** [comment]
> Spark 3.1 truncates timestamp to date?

@cloud-fan Yep. Spark 3.1 as well as master when `spark.sql.legacy.interval.enabled` is set to `true` casts the result to the type of the left parameter.

> I think the new behavior makes more sense

It depends how you get the interval. For instance,

Spark 3.1:
```sql
spark-sql> select date'now' + (date'now' - date'yesterday');
2021-04-15
```

After the changes:
```sql
spark-sql> select date'now' + (date'now' - date'yesterday');
2021-04-15 00:00:00
```
As you can see, Spark 3.1 returns a date as the result which also makes sense.

> shall we add an item in the migration guide?

Adding ANSI day-time interval to a date wasn't supported in Spark 3.1. Actually, the behavior change is related to subtraction of dates/timestamps which has been already documented in the SQL migration guide:
```
  - In Spark 3.2, the dates subtraction expression such as `date1 - date2` returns values of `DayTimeIntervalType`. In Spark 3.1 and earlier, the returned type is `CalendarIntervalType`. To restore the behavior before Spark 3.2, you can set `spark.sql.legacy.interval.enabled` to `true`.

  - In Spark 3.2, the timestamps subtraction expression such as `timestamp '2021-03-31 23:48:00' - timestamp '2021-01-01 00:00:00'` returns values of `DayTimeIntervalType`. In Spark 3.1 and earlier, the type of the same expression is `CalendarIntervalType`. To restore the behavior before Spark 3.2, you can set `spark.sql.legacy.interval.enabled` to `true`.
```

**Review by cloud-fan — APPROVED** [review]

**Comment by MaxGekk** [comment]
All tests passed:
Screenshot 2021-04-14 at 19 26 12
Merging to master. Thank you @cloud-fan for your review.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala [hunk 1]
```diff
@@ -345,6 +345,8 @@ class Analyzer(override val catalogManager: CatalogManager)
     override def apply(plan: LogicalPlan): LogicalPlan = plan.resolveOperatorsUp {
       case p: LogicalPlan => p.transformExpressionsUp {
         case a @ Add(l, r, f) if a.childrenResolved => (l.dataType, r.dataType) match {
+          case (DateType, DayTimeIntervalType) => TimeAdd(Cast(l, TimestampType), r)
+          case (DayTimeIntervalType, DateType) => TimeAdd(Cast(r, TimestampType), l)
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala:349]: I doubt this. For old interval types, `Date + Interval` always returns Date:
```
case (DateType, CalendarIntervalType) => DateAddInterval(l, r, ansiEnabled = f)
case (CalendarIntervalType, DateType) => DateAddInterval(r, l, ansiEnabled = f)
```

I think we should do this for new interval types as well, and fail at runtime if the interval has hour, minute, second fields.

**Pull request closed without merging**